### ISSUE TITLE
Spike - Investigate requests impacting the (SLO) of the API

### ISSUE THREAD

**Issue by adberglund** [project-koku/koku#1707]
## User Story
As an a customer, I want API responses to not return with internal errors and to return quickly, so that my expected data is returned in a responsive timeframe.

## Impacts
API

## Assumptions
- Investigator has access to Preproduction and Production Kibana
-

## Implementation Details
- Investigate requests causing 5xxs in Preproduction and open bugs with any associated findings:
https://kibana-kibana.5a9f.insights-dev.openshiftapps.com/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),index:d2da0310-7048-11e9-a501-7998195d35a8,interval:auto,query:(language:lucene,query:'gateway.application:%20%22cost-management%22%20AND%20gateway.status:%20%5B500%20TO%20599%5D'),sort:!('@timestamp',desc))
*Search Query:* `gateway.application: "cost-management" AND gateway.status: [500 TO 599]`

- Investigate requests causing 5xxs in Production and open bugs with any associated findings if not already covered or fixed in preproduction or by Sentry:
https://kibana-kibana.1b13.insights.openshiftapps.com/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),filters:!(),index:'0a2575c0-65f9-11e9-ba88-097c6c9f6850',interval:auto,query:(language:lucene,query:'gateway.application:%20%22cost-management%22%20AND%20gateway.status:%20%5B500%20TO%20599%5D'),sort:!('@timestamp',desc))
*Search Query:* `gateway.application: "cost-management" AND gateway.status: [500 TO 599]`

- Investigate requests impacting latency (taking longer than 2 seconds) in Preproduction and open bugs with any associated findings:
https://kibana-kibana.5a9f.insights-dev.openshiftapps.com/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),index:d2da0310-7048-11e9-a501-7998195d35a8,interval:auto,query:(language:lucene,query:'gateway.application:%20%22cost-management%22%20AND%20gateway.time_taken:%20%5B2%20TO%20999999999999999%5D'),sort:!('@timestamp',desc))
*Search Query:* `gateway.application: "cost-management" AND gateway.time_taken: [2 TO 999999999999999]`

- Investigate requests impacting latency (taking longer than 2 seconds) in Production and open bugs with any associated findings:
https://kibana-kibana.1b13.insights.openshiftapps.com/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),filters:!(),index:%270a2575c0-65f9-11e9-ba88-097c6c9f6850%27,interval:auto,query:(language:lucene,query:%27gateway.application:%20%22cost-management%22%20AND%20gateway.time_taken:%20%5B2%20TO%20999999999999999%5D%27),sort:!(%27@timestamp%27,desc))
*Search Query:* `gateway.application: "cost-management" AND gateway.time_taken: [2 TO 999999999999999]`

## Acceptance Criteria

- [ ] Verify that bugs are created to cover the error and latency related issues

**Comment by adberglund** [comment]
@chambridge to fix links to queries

**Comment by nbon12** [comment]
Sprint Planning Meeting Minutes:

Andrew: some of this work is on-going, we already tackled parameter gating, instance counts on the AWS instance API, doing the provider API fixes. The query link listed in this issue should open up to a query that's checking for 5XX errors in Kibana. The latency queries in pre-prod and prod show requests taking longer than 2 seconds.

**Comment by chambridge** [comment]
Provided search query in case the URL launch doesn't work, although I think it only breaks if launched via ZenHub, it seems to be fine from GitHub.

FYI - @adberglund

**Comment by blentz** [comment]
I looked into this. 

I did not find any 5xx errors that I could attribute to bugs or improper app behavior.

I have compiled a list of the worst performing URLs:
https://docs.google.com/spreadsheets/d/1we3RvsnrgmgWAzqGSh9P_0SK2jKtWBaBZigPXi3N55o/edit?usp=sharing

The things that stand out to me are:
- The Providers endpoint probably needs to be made asynchronous, so that the underlying DB work isn't blocking returning an HTTP response to the user.
- The Cost Models endpoint likely also needs to be async.
- If Providers and Cost Models need to be async, we may want to consider making all of the basic CRUD calls async (i.e. everything that isn't a reporting endpoint).
- The reporting endpoint slowness is something we're already addressing from several directions. I don't have any additional insight there.
